src/cata_ui.cpp

```diff
@@ -0,0 +1,442 @@
+#include "cata_ui.h"
+
+#include "catacharset.h"
+#include "output.h"
+#include "input.h"
+
+#include <cmath>
+#include <array>
+#include <unordered_set>
+
+ui_rect::ui_rect( size_t size_x, size_t size_y, int x, int y ) : size_x( size_x ), size_y( size_y ), x( x ), y( y )
+{
+}
+
+ui_window::ui_window( const ui_rect &rect, ui_anchor anchor ) : ui_element( rect, anchor ),
+                    global_x( get_ax() ), global_y( get_ay() ), win( newwin( rect.size_y, rect.size_x, global_y, global_x ) )
+{
+}
+
+ui_window::ui_window( size_t size_x, size_t size_y, int x, int y, ui_anchor anchor ) : ui_window( ui_rect( size_x, size_y, x, y ), anchor )
+{
+}
+
+ui_window::~ui_window()
+{
+    for( auto c : children ){
+        delete c;
+    }
+
+    delwin( win );
+}
+
+void ui_window::draw( std::vector<WINDOW *> &render_batch )
+{
+    werase( win );
+    draw( win );
+    render_batch.push_back( win );
+    draw_children( render_batch );
+}
+
+void ui_window::draw()
+{
+    std::vector<WINDOW *> batch;
+    draw( batch );
+
+    std::unordered_set<WINDOW *> had;
```

Comment:
If you're using the unordered set here, why not use it as type of the `render_batch` parameter? That way the container does not even contain a duplicate ever. You have to create the set anyway (here or as `render_batch` parameter).